third_party/amd/lib/TritonAMDGPUTransforms/inThreadTranspose.cpp

```diff
@@ -0,0 +1,213 @@
+  auto v = op->getOperand(0);
+  auto prevOp = v.getDefiningOp();
+  if (isa<RegionBranchOpInterface>(prevOp)) {
+    LDBG("Dealing with scf blocks");
+    auto idx = cast<OpResult>(v).getResultNumber();
+    llvm::SmallVector<scf::YieldOp> yieldOps;
+    prevOp->walk([&](Operation *op) {
+      if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
+        yieldOps.push_back(yieldOp);
+      }
+    });
+
+    for (auto yieldOp : yieldOps) {
+      auto maybeLoadOp = yieldOp.getOperand(idx).getDefiningOp();
+      if (isa<tt::LoadOp>(maybeLoadOp))
+        ret.push_back(maybeLoadOp);
+    }
+  } else {
+    // regular case
+    LDBG("Regular cases");
+    assert(isa<tt::LoadOp>(prevOp) &&
+           "only load->convert_layout->dot is supported.");
+    ret.push_back(prevOp);
+  }
+  return ret;
+}
+
+bool convertToThreadRaked(Value operand) {
+  auto opTensorTy = cast<RankedTensorType>(operand.getType());
+  auto opEnc = opTensorTy.getEncoding();
+  auto opDotOpEnc = dyn_cast<ttg::DotOperandEncodingAttr>(opEnc);
+  if (!opDotOpEnc)
+    return false;
+  if (!isa<ttg::AMDMfmaEncodingAttr>(opDotOpEnc.getParent())) {
+    LDBG("Operand's parent encoding is not MFMA");
+    return false;
+  }
+  auto cvtOperand = operand.getDefiningOp()->getOperand(0);
+  auto cvtOperandEnc =
+      cast<RankedTensorType>(cvtOperand.getType()).getEncoding();
+  auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(cvtOperandEnc);
+  if (!blockedEnc)
+    return false;
+  int kDimNum = opDotOpEnc.getOpIdx() == 0 ? 1 : 0;
+  auto order = blockedEnc.getOrder();
+  if (order[0] != kDimNum) {
+    return true;
+  }
+
+  return false;
+}
+
+ttg::BlockedEncodingAttr getThreadRakedBlockedEnc(Value operand,
+                                                  ModuleOp &mod) {
+  // get the K dim according to dotOp operand's index
+  auto tensorTy = cast<RankedTensorType>(operand.getType());
+  auto shape = tensorTy.getShape();
+  auto opEnc = tensorTy.getEncoding();
+  auto opDotOpEnc = dyn_cast<ttg::DotOperandEncodingAttr>(opEnc);
+  int kDimNum = opDotOpEnc.getOpIdx() == 0 ? 1 : 0;
+  // get the current blocked encoding
+  auto cvtOperand = operand.getDefiningOp()->getOperand(0);
+  auto cvtOperandEnc =
+      cast<RankedTensorType>(cvtOperand.getType()).getEncoding();
+  auto blockedEnc = dyn_cast<ttg::BlockedEncodingAttr>(cvtOperandEnc);
+  // compute the sizePerThread for the new encoding
+  auto sizePerThread = blockedEnc.getSizePerThread();
+  auto elemsPerIter = product(sizePerThread);
+  auto elemsTotal = blockedEnc.getTotalElemsPerThread(shape, tensorTy);
+  // we need to know how many iteration each thread will load
+  LDBG("elemsPerIter = " << elemsPerIter << "; elemsTotal = " << elemsTotal);
+  auto numMaxIters = elemsTotal / elemsPerIter;
+  auto bitwidth = tensorTy.getElementType().getIntOrFloatBitWidth();
+  // Current the widest is set to ds_write_b64
+  auto newKOuterDim = std::min(numMaxIters, 64 / bitwidth);
```

Comment:
Why can't we do ds_write_b128?

Reply:
tbh I just fixed this 64bits initially and never changed it afterwards. I can experiment on this later when I have all the PRs in to see if relaxing it to 128bits is better, and change accordingly